File: workers/llm-worker/src/scripts/generate-benchmark-share-card.ts
import { mkdir, readdir, readFile, writeFile } from "node:fs/promises";
import { join, resolve } from "node:path";

type BenchmarkScorecard = {
	generatedAt: string;
	repoBenchmarks: {
		fixtures: number;
		cases: number;
		bucketAccuracy: number;
		verdictAccuracy: number;
		fixtureBreakdown: Array<{
			id: string;
			name: string;
			description: string;
			cases: number;
			bucketAccuracy: number;
			verdictAccuracy: number;
		}>;
	};
	auditorReviewSet: {
		cases: number;
		headlineAccuracy: number;
		bucketAccuracy: number;
		verdictAccuracy: number;
		focusAreas: Array<{ tag: string; count: number }>;
		caseHighlights: Array<{
			id: string;
			title: string;
			detector: string;
			expectedBucket: string;
			shouldHeadline: boolean;
		}>;
	};
	auditIntelligence: {
		artifacts: number;
		reportFindings: number;
		needsReview: number;
		researchNotes: number;
		provenanceCoverage: number;
		valueFlowCoverage: number;
		dimensionalCoverage: number;
		dimensionalMismatchCoverage: number;
		topImprovementSignals: Array<{ key: string; count: number }>;
		topDimensionalMismatchKinds: Array<{ kind: string; count: number }>;
		recentTargets: Array<{
			projectId: string;
			projectTitle: string;
			jobId: string;
		}>;
	};
};

const resultsDir = resolve(process.cwd(), "benchmarks", "results");
const shareDir = resolve(process.cwd(), "benchmarks", "share");

async function getLatestScorecardPath() {
	const entries = await readdir(resultsDir, { withFileTypes: true });
	const files = entries
		.filter((entry) => entry.isFile() && entry.name.endsWith(".json"))
		.map((entry) => entry.name)
		.sort()
		.reverse();
	if (files.length === 0) {
		throw new Error(`No benchmark scorecards found in ${resultsDir}`);
	}
	return join(resultsDir, files[0]);
}

function pct(value: number) {
	return `${(value * 100).toFixed(1)}%`;
}

function escapeXml(value: string) {
	return String(value)
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&apos;");
}

function renderMetricCard(
	x: number,
	y: number,
	label: string,
	value: string,
	subvalue: string,
	fill: string,
) {
	return `
		<g transform="translate(${x} ${y})">
			<rect width="340" height="122" rx="10" fill="${fill}" stroke="#27272A" />
			<text x="22" y="34" fill="#A1A1AA" font-size="16" font-family="Inter, Arial, sans-serif" font-weight="700">${escapeXml(label.toUpperCase())}</text>
			<text x="22" y="76" fill="#FAFAFA" font-size="40" font-family="Inter, Arial, sans-serif" font-weight="800">${escapeXml(value)}</text>
			<text x="22" y="101" fill="#D4D4D8" font-size="17" font-family="Inter, Arial, sans-serif">${escapeXml(subvalue)}</text>
		</g>
	`;
}

function renderTextLines(
	x: number,
	startY: number,
	lines: string[],
	fontSize: number,
	lineHeight: number,
	color: string,
	fontWeight = "400",
) {
	return lines
		.map(
			(line, index) => `<text x="${x}" y="${startY + index * lineHeight}" fill="${color}" font-size="${fontSize}" font-family="Inter, Arial, sans-serif" font-weight="${fontWeight}">${escapeXml(line)}</text>`,
		)
		.join("\n");
}

function buildSvg(scorecard: BenchmarkScorecard) {
	const topSignal =
		scorecard.auditIntelligence.topImprovementSignals[0]?.key || "none";
	const topMismatch =
		scorecard.auditIntelligence.topDimensionalMismatchKinds[0]?.kind || "none";
	const corpusLines = scorecard.repoBenchmarks.fixtureBreakdown
		.slice(0, 4)
		.map((fixture) => `${fixture.name} (${fixture.cases} cases)`);
	const focusLines = scorecard.auditorReviewSet.focusAreas
		.slice(0, 4)
		.map((area) => `${area.tag} (${area.count})`);
	const methodologyLines = [
		"Evidence layers",
		"1. curated repo benchmark fixtures",
		"2. auditor-aligned reviewset expectations",
		"3. anonymized production telemetry",
	];
	const telemetryLines = [
		`${scorecard.auditIntelligence.artifacts} local audit artifacts`,
		`${scorecard.auditIntelligence.reportFindings} report findings / ${scorecard.auditIntelligence.needsReview} review findings`,
		`top signal: ${topSignal}`,
		`top mismatch: ${topMismatch}`,
	];

	return `<?xml version="1.0" encoding="UTF-8"?>
<svg width="1600" height="900" viewBox="0 0 1600 900" fill="none" xmlns="http://www.w3.org/2000/svg">
	<rect width="1600" height="900" fill="#000000"/>
	<line x1="80" y1="84" x2="1520" y2="84" stroke="#27272A"/>
	<line x1="80" y1="822" x2="1520" y2="822" stroke="#27272A"/>
	<text x="80" y="70" fill="#FAFAFA" font-size="18" font-family="Inter, Arial, sans-serif" font-weight="700">SENTINELAUDIT BENCHMARK SNAPSHOT</text>
	<text x="80" y="155" fill="#FAFAFA" font-size="60" font-family="Georgia, Times New Roman, serif" font-weight="700">Security Workflow Benchmarking</text>
	<text x="80" y="198" fill="#A1A1AA" font-size="22" font-family="Inter, Arial, sans-serif">Generated ${escapeXml(scorecard.generatedAt.slice(0, 10))} from deterministic benchmark suites and local audit telemetry.</text>

	${renderMetricCard(
		80,
		248,
		"Repo Benchmarks",
		pct(scorecard.repoBenchmarks.bucketAccuracy),
		`${scorecard.repoBenchmarks.fixtures} fixtures - ${scorecard.repoBenchmarks.cases} cases`,
		"#0A0A0A",
	)}
	${renderMetricCard(
		450,
		248,
		"Auditor Review Set",
		pct(scorecard.auditorReviewSet.headlineAccuracy),
		`${scorecard.auditorReviewSet.cases} reviewed cases`,
		"#0A0A0A",
	)}
	${renderMetricCard(
		820,
		248,
		"Provenance Coverage",
		pct(scorecard.auditIntelligence.provenanceCoverage),
		`${scorecard.auditIntelligence.artifacts} local artifacts`,
		"#0A0A0A",
	)}
	${renderMetricCard(
		1190,
		248,
		"Value-Flow Coverage",
		pct(scorecard.auditIntelligence.valueFlowCoverage),
		`${scorecard.auditIntelligence.reportFindings} report - ${scorecard.auditIntelligence.needsReview} review`,
		"#0A0A0A",
	)}

	<g transform="translate(80 430)">
		<rect width="700" height="338" rx="14" fill="#0A0A0A" stroke="#27272A" />
		<text x="30" y="44" fill="#FAFAFA" font-size="28" font-family="Inter, Arial, sans-serif" font-weight="800">Public benchmark corpus</text>
		<text x="30" y="77" fill="#A1A1AA" font-size="19" font-family="Inter, Arial, sans-serif">Named fixtures and review focus areas that can be inspected directly.</text>
		${renderTextLines(30, 130, corpusLines, 21, 34, "#FAFAFA", "600")}
		<text x="30" y="272" fill="#A1A1AA" font-size="18" font-family="Inter, Arial, sans-serif" font-weight="700">Auditor review focus</text>
		${renderTextLines(30, 304, focusLines, 20, 30, "#D4D4D8", "400")}
	</g>

	<g transform="translate(820 430)">
		<rect width="700" height="338" rx="14" fill="#0A0A0A" stroke="#27272A" />
		<text x="34" y="44" fill="#FAFAFA" font-size="28" font-family="Inter, Arial, sans-serif" font-weight="800">Telemetry and method</text>
		${renderTextLines(34, 92, telemetryLines, 20, 30, "#FAFAFA", "500")}
		<text x="34" y="238" fill="#A1A1AA" font-size="18" font-family="Inter, Arial, sans-serif" font-weight="700">Methodology</text>
		${renderTextLines(34, 272, methodologyLines, 20, 28, "#D4D4D8")}
	</g>

	<text x="80" y="852" fill="#71717A" font-size="18" font-family="Inter, Arial, sans-serif">Public benchmark corpus is inspectable. Production telemetry is anonymized by default.</text>
</svg>`;
}

function buildMarkdown(scorecard: BenchmarkScorecard) {
	const topSignal =
		scorecard.auditIntelligence.topImprovementSignals[0]?.key || "none";
	const topMismatch =
		scorecard.auditIntelligence.topDimensionalMismatchKinds[0]?.kind || "none";

	return `# SentinelAudit Benchmark Snapshot

- Generated: ${scorecard.generatedAt}
- Repo benchmarks: ${pct(scorecard.repoBenchmarks.bucketAccuracy)} bucket accuracy across ${scorecard.repoBenchmarks.fixtures} fixtures / ${scorecard.repoBenchmarks.cases} cases
- Auditor review set: ${pct(scorecard.auditorReviewSet.headlineAccuracy)} headline accuracy across ${scorecard.auditorReviewSet.cases} cases
- Audit intelligence artifacts: ${scorecard.auditIntelligence.artifacts}
- Provenance coverage: ${pct(scorecard.auditIntelligence.provenanceCoverage)}
- Value-flow coverage: ${pct(scorecard.auditIntelligence.valueFlowCoverage)}
- Dimensional coverage: ${pct(scorecard.auditIntelligence.dimensionalCoverage)}
- Dimensional mismatch coverage: ${pct(scorecard.auditIntelligence.dimensionalMismatchCoverage)}
- Top improvement signal: ${topSignal}
- Top mismatch kind: ${topMismatch}

Public benchmark corpus:
${scorecard.repoBenchmarks.fixtureBreakdown
	.slice(0, 6)
	.map((fixture) => `- ${fixture.name}: ${fixture.cases} cases`)
	.join("\n")}

Auditor review focus:
${scorecard.auditorReviewSet.focusAreas
	.slice(0, 6)
	.map((area) => `- ${area.tag}: ${area.count}`)
	.join("\n")}

Recent audited targets (local telemetry):
${scorecard.auditIntelligence.recentTargets.length > 0
	? scorecard.auditIntelligence.recentTargets
			.slice(0, 5)
			.map(
				(target) =>
					`- ${target.projectTitle} (${target.projectId}) – job ${target.jobId}`,
			)
			.join("\n")
	: "- none"}

Methodology:
- curated repo benchmark fixtures
- auditor-aligned reviewset expectations
- anonymized audit-intelligence telemetry from real audit runs
`;
}

async function main() {
	const latestScorecardPath = await getLatestScorecardPath();
	const scorecard = JSON.parse(
		await readFile(latestScorecardPath, "utf8"),
	) as BenchmarkScorecard;

	await mkdir(shareDir, { recursive: true });
	const stem = latestScorecardPath.split("\\").pop()?.replace(/\.json$/i, "") || "latest";
	const svgPath = join(shareDir, `${stem}.svg`);
	const mdPath = join(shareDir, `${stem}.md`);

	await writeFile(svgPath, buildSvg(scorecard), "utf8");
	await writeFile(mdPath, buildMarkdown(scorecard), "utf8");

	console.log("=== SentinelAudit Benchmark Share Assets ===");
	console.log(`Source scorecard: ${latestScorecardPath}`);
	console.log(`SVG card: ${svgPath}`);
	console.log(`Markdown summary: ${mdPath}`);
}

main().catch((error) => {
	console.error("[benchmark-share-card] Failed:", error);
	process.exitCode = 1;
});

File: workers/llm-worker/src/scripts/generate_benchmark_share_card.py
from __future__ import annotations

import json
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont


ROOT = Path(__file__).resolve().parents[2]
RESULTS_DIR = ROOT / "benchmarks" / "results"
SHARE_DIR = ROOT / "benchmarks" / "share"


def pct(value: float) -> str:
    return f"{value * 100:.1f}%"


def load_latest_scorecard() -> tuple[Path, dict]:
    files = sorted(RESULTS_DIR.glob("*.json"), reverse=True)
    if not files:
        raise RuntimeError(f"No benchmark scorecards found in {RESULTS_DIR}")
    latest = files[0]
    return latest, json.loads(latest.read_text(encoding="utf-8"))


def load_font(size: int, bold: bool = False) -> ImageFont.FreeTypeFont | ImageFont.ImageFont:
    candidates = [
        "C:/Windows/Fonts/georgiab.ttf" if bold else "C:/Windows/Fonts/georgia.ttf",
        "C:/Windows/Fonts/arialbd.ttf" if bold else "C:/Windows/Fonts/arial.ttf",
    ]
    for path in candidates:
        try:
            return ImageFont.truetype(path, size=size)
        except OSError:
            continue
    return ImageFont.load_default()


def wrap_text(draw: ImageDraw.ImageDraw, text: str, font: ImageFont.ImageFont, max_width: int) -> list[str]:
    words = text.split()
    if not words:
        return [""]
    lines: list[str] = []
    current = words[0]
    for word in words[1:]:
        candidate = f"{current} {word}"
        if draw.textbbox((0, 0), candidate, font=font)[2] <= max_width:
            current = candidate
        else:
            lines.append(current)
            current = word
    lines.append(current)
    return lines


def draw_metric(
    draw: ImageDraw.ImageDraw,
    box: tuple[int, int, int, int],
    label: str,
    value: str,
    sub: str,
):
    x1, y1, x2, y2 = box
    draw.rounded_rectangle(box, radius=12, fill="#0A0A0A", outline="#27272A", width=2)
    draw.text((x1 + 22, y1 + 18), label.upper(), fill="#A1A1AA", font=FONT_LABEL)
    draw.text((x1 + 22, y1 + 48), value, fill="#FAFAFA", font=FONT_VALUE)
    draw.text((x1 + 22, y1 + 94), sub, fill="#D4D4D8", font=FONT_TEXT)


FONT_TITLE = load_font(44, bold=True)
FONT_SUBTITLE = load_font(18)
FONT_LABEL = load_font(18, bold=True)
FONT_VALUE = load_font(34, bold=True)
FONT_TEXT = load_font(18)
FONT_BODY = load_font(22)


def main():
    latest_path, scorecard = load_latest_scorecard()
    SHARE_DIR.mkdir(parents=True, exist_ok=True)

    width, height = 1600, 900
    image = Image.new("RGB", (width, height), "#000000")
    draw = ImageDraw.Draw(image)
    draw.line((80, 84, 1520, 84), fill="#27272A", width=1)
    draw.line((80, 822, 1520, 822), fill="#27272A", width=1)

    draw.text((80, 52), "SENTINELAUDIT BENCHMARK SNAPSHOT", fill="#FAFAFA", font=FONT_LABEL)
    draw.text((80, 108), "Security Workflow Benchmarking", fill="#FAFAFA", font=FONT_TITLE)
    draw.text(
        (80, 170),
        f"Generated {scorecard['generatedAt'][:10]} from deterministic benchmark suites and local audit telemetry.",
        fill="#A1A1AA",
        font=FONT_SUBTITLE,
    )

    repo = scorecard["repoBenchmarks"]
    auditor = scorecard["auditorReviewSet"]
    intelligence = scorecard["auditIntelligence"]

    draw_metric(
        draw,
        (80, 248, 420, 370),
        "Repo Benchmarks",
        pct(repo["bucketAccuracy"]),
        f"{repo['fixtures']} fixtures - {repo['cases']} cases",
    )
    draw_metric(
        draw,
        (450, 248, 790, 370),
        "Auditor Review Set",
        pct(auditor["headlineAccuracy"]),
        f"{auditor['cases']} reviewed cases",
    )
    draw_metric(
        draw,
        (820, 248, 1160, 370),
        "Provenance Coverage",
        pct(intelligence["provenanceCoverage"]),
        f"{intelligence['artifacts']} local artifacts",
    )
    draw_metric(
        draw,
        (1190, 248, 1530, 370),
        "Value-Flow Coverage",
        pct(intelligence["valueFlowCoverage"]),
        f"{intelligence['reportFindings']} report - {intelligence['needsReview']} review",
    )

    draw.rounded_rectangle((80, 430, 780, 780), radius=14, fill="#0A0A0A", outline="#27272A", width=2)
    draw.text((110, 464), "Public benchmark corpus", fill="#FAFAFA", font=FONT_LABEL)
    draw.text((110, 506), "Named fixtures and review focus areas that can be inspected directly.", fill="#A1A1AA", font=FONT_TEXT)
    corpus_lines = [
        f"{fixture['name']} ({fixture['cases']} cases)"
        for fixture in repo["fixtureBreakdown"][:4]
    ]
    y = 556
    for line in corpus_lines:
        draw.text((110, y), line, fill="#FAFAFA", font=FONT_BODY)
        y += 34
    draw.text((110, 690), "Auditor review focus", fill="#A1A1AA", font=FONT_TEXT)
    focus_lines = [f"{area['tag']} ({area['count']})" for area in auditor["focusAreas"][:4]]
    y = 724
    for line in focus_lines:
        draw.text((110, y), line, fill="#D4D4D8", font=FONT_TEXT)
        y += 28

    draw.rounded_rectangle((820, 430, 1530, 780), radius=14, fill="#0A0A0A", outline="#27272A", width=2)
    draw.text((850, 464), "Telemetry and method", fill="#FAFAFA", font=FONT_LABEL)
    top_signal = intelligence["topImprovementSignals"][0]["key"] if intelligence["topImprovementSignals"] else "none"
    top_mismatch = (
        intelligence["topDimensionalMismatchKinds"][0]["kind"]
        if intelligence["topDimensionalMismatchKinds"]
        else "none"
    )
    telemetry_lines = [
        f"{intelligence['artifacts']} local audit artifacts",
        f"{intelligence['reportFindings']} report findings / {intelligence['needsReview']} review findings",
        f"top signal: {top_signal}",
        f"top mismatch: {top_mismatch}",
    ]
    y = 516
    for line in telemetry_lines:
        draw.text((850, y), line, fill="#FAFAFA", font=FONT_TEXT)
        y += 30
    draw.text((850, 648), "Methodology", fill="#A1A1AA", font=FONT_TEXT)
    method_lines = [
        "Evidence layers",
        "1. curated repo benchmark fixtures",
        "2. auditor-aligned reviewset expectations",
        "3. anonymized production telemetry",
    ]
    y = 682
    for line in method_lines:
        draw.text((850, y), line, fill="#D4D4D8", font=FONT_TEXT)
        y += 28

    draw.text(
        (80, 844),
        "Public benchmark corpus is inspectable. Production telemetry is anonymized by default.",
        fill="#71717A",
        font=FONT_TEXT,
    )

    stem = latest_path.stem
    out_path = SHARE_DIR / f"{stem}.png"
    image.save(out_path, format="PNG")
    print("=== SentinelAudit Benchmark PNG ===")
    print(f"Source scorecard: {latest_path}")
    print(f"PNG card: {out_path}")


if __name__ == "__main__":
    main()
